GliaWidgets/SecureConversations/SecureConversations.Availability.swift

```diff
@@ -0,0 +1,50 @@
+import Foundation
+
+extension SecureConversations {
+    struct Availability {
+        typealias CompletionResult = (Result<Bool, CoreSdkClient.SalemoveError>) -> Void
+        var environment: Environment
+
+        func checkSecureConversationsAvailability(completion: @escaping CompletionResult) {
+            environment.listQueues { queues, error in
+                if let error = error {
+                    completion(.failure(error))
+                    return
+                }
+
+                self.checkQueues(queues, completion: completion)
+            }
+        }
+
+        private func checkQueues(
+            _ queues: [CoreSdkClient.Queue]?,
+            completion: (Result<Bool, CoreSdkClient.SalemoveError>) -> Void
+        ) {
+            guard let queues = queues, isSecureConversationsAvailable(in: queues) else {
+                completion(.success(false))
+
+                return
+            }
+
+            completion(.success(true))
+        }
+
+        private func isSecureConversationsAvailable(
+            in queues: [CoreSdkClient.Queue]
+        ) -> Bool {
+            let filteredQueues = queues
+                .filter { self.environment.queueIds.contains($0.id) }
+                .filter { $0.state.status != .closed }
+                .filter { $0.state.media.contains(CoreSdkClient.MediaType.messaging) }
+
+            return !filteredQueues.isEmpty
```

Comment:
in sense of optimization, I'd like to propose using one `contains(where:)` or `first(where:)` instead of three filters. What do you think, does it make sense?

Reply:
Makes complete sense, I would guess that maybe Swift when compiling would optimize it into just making one, but I am not sure, so fbb0d8ef7636a0daeeaf95d166b4084fca2f42f9. In Elixir you can make the first two calls lazy using the `Stream` module and then on the last one you use the `Map` module and the three would be applied at the same time, so that you don't have to do && all the time. But I don't know Swift has something like that.